### DISCUSSION TITLE
Variable size property

### DISCUSSION THREAD

**Discussion by ErfanBabakhani** [esa/asn1scc#318] [Q&A]
Hello
I have a question. Suppose that I have a field named lenght1 and other field named length2 and both of them declared in ACN.
I also have a field named Message which declared in ASN and now I want to set its size according to  lenght2 and also length1
As I see your compiler asn1scc has ability to set the size of a field that declared in ASN according to specific field declared in ACN, but how can I use two fields. In this case suppose I want to set the size of the Message as length1 - length2 (assume I know length1 - length2 is always nonnegative INTEGER)

I mean somthing like :
```
--ASN DEF
Erf DEFINITIONS AUTOMATIC TAGS ::= BEGIN
    Packet ::=SEQUENCE {
        mese	Mes
    }
    Mes ::=OCTET STRING (SIZE(1..5))
    
END

--ACN DEF
Erf DEFINITIONS ::= BEGIN
    Packet [] {
        len1 INTEGER [size 3 , encoding pos-int], 
        len2 INTEGER [size 3 , encoding pos-int], 
        mese	[size (len2 - len1)]  --test.acn:5:14: error: no viable alternative at input 'size'
    }
    
    
END
```

This situation occurred when I want to calculate the size of specific field from entire PDU body size and other variable size (also some const value)

**Comment by ErfanBabakhani** [comment]
Because of this lack I have to use optional paramethers and control their existence to handel the size I want reach

**Comment by maxime-esa** [comment]
Correct me if I am wrong but I think there is a theoretical impossibility with this specific example, if we assume that the encoder is deterministic.

Concretely what values will you place in `len1` and `len2` to achieve what you want (`size = len2 - len1`) ?

If the size is e.g. 4, there are multiple possibilities (`7 - 3`, `5 - 1`, `4 - 0`, `6 - 2` with the 3 bits encoding you specified).

**Reply by ErfanBabakhani** [reply]
```
Example DEFINITIONS AUTOMATIC TAGS::=BEGIN
    LenOfMessage ::= INTEGER(0..15) 
    LenOfPacketEntire ::= INTEGER(0..15)
    Message ::= OCTET STRING(SIZE(0..15)) --vatiable size  , size equal to LenOfMessage value
    VarWithComplexSize ::=OCTET STRING (SIZE(0..15)) --variable size , size equal with (LenOfPacketEntire - LenOfMessage value - 1)
    Packet ::= SEQUENCE {
        lenOfMessagee LenOfMessage,
        lenOfPacketEntiree LenOfPacketEntire,
        messagee Message ,
        varWithComplexSizee VarWithComplexSize
    }

END
```
This may be a better example
I mean supppose that I want have message with variable size and I also consider specefic place for it <No problem and Ok>
But I have also other field like VarWithComplexSize that has size equal to specefic field name LenOfPacketEntire - LenOfMessage - 1 (1 because lenOfPacketEntiree and lenOfMessagee size sum up is 1 byte)

Suppose situation I have field with variable size but I can not put it's size in a single specefic field insead I calculate it with other fields

**Comment by maxime-esa** [comment]
This is relevant to the discussion, please check it out: https://github.com/maxime-esa/asn1scc/blob/master/Docs/deduced-size.md

**Reply by ErfanBabakhani** [reply]
I read it but the problem is I need entire size of my pdu before decoding it
But in my case I send it with in my PDU 

I try to solve my problem with this way :  
```
Example DEFINITIONS AUTOMATIC TAGS::=BEGIN
    LenOfMessage ::= INTEGER(0..7) -- suppose with size 4 bit <set it in ACN>
    LenOfPacketEntire ::= INTEGER(0..15) 
    Packet ::= SEQUENCE {
        lenOfMessagee LenOfMessage,
        lenOfPacketEntiree LenOfPacketEntire,
        message3 OCTET STRING (SIZE(4)) OPTIONAL, --we should handle message3 to message1 existence in the way th entire size of them be equal to lenOfMessage value
        message2 OCTET STRING (SIZE(2)) OPTIONAL,
        message1 OCTET STRING (SIZE(1)) OPTIONAL,
        varWithComplexSize4   OCTET STRING (SIZE(8)) OPTIONAL, --we should handle varWithComplexSize4 to varWithComplexSize1 existence in the way th entire size of them be equal to (lenOfPacketEntiree - lenOfMessage) value
        varWithComplexSize3   OCTET STRING (SIZE(4)) OPTIONAL,
        varWithComplexSize2   OCTET STRING (SIZE(2)) OPTIONAL,
        varWithComplexSize1   OCTET STRING (SIZE(1)) OPTIONAL
    }

END

--ACN
Example DEFINITIONS ::=BEGIN
    LenOfMessage [size 4 , encoding pos-int]
    LenOfPacketEntire [size 4 , encoding pos-int]
    Packet [] {
        lenOfMessagee [],
        lenOfPacketEntiree [],
        message3 [present-when (lenOfMessagee >=4)],
        message2 [present-when (lenOfMessagee %4) >=2],
        message1 [present-when (lenOfMessagee %2) >=1],
        varWithComplexSize4  [present-when (lenOfPacketEntiree - lenOfMessagee - 1) >=8 ],
        varWithComplexSize3  [present-when (lenOfPacketEntiree - lenOfMessagee - 1 ) %8 >=4], 
        varWithComplexSize2  [present-when (lenOfPacketEntiree - lenOfMessagee - 1 ) %4 >=2], 
        varWithComplexSize1  [present-when (lenOfPacketEntiree - lenOfMessagee - 1) %2 >=1]
    }

END
```

I want to say if you add some extra ability to your ACN compiler it can make it more stronger and you have done it already for optional fields (controlling their existence with multiple manddetory fields and simple logic operators like - + * % ..)

**Reply by ErfanBabakhani** [reply]
Thank you for your powerfull free asn1scc compiler . It is really usefull and make our life easier

**Reply by ErfanBabakhani** [reply]
I tried the document example you give


MySequenceOf ::= SEQUENCE (SIZE (0..100)) OF SOME-TYPE 
MySequenceOf [deduced-size]

This Error ocuerd
asn1scc -c -ACN -XER -equal -o deduced-size/ test.asn test.acn 
test.acn:4:12: error: extraneous input 'deduced-size' expecting R_SBRACKET

**Reply by maxime-esa** [reply]
Yes, this doc is for the moment just a spec to address the issue of "deduced" size fields, on cases that are a bit different from yours but still related. It's not implemented yet.

**Reply by ErfanBabakhani** [reply]
Hy,
I have a question 
Is it possible to have a variable size field with an ACN field for its size determining but also have some offset I mean if I have a arraySize field in ACN and also I have a variable size field varSize in ASN then determine its size like that varSize[size arraySize - <constant value as it's offset>]?

I also want to make an issue:
Suppose that  I have a field named SizeOfHeader that Exist in my Packet Header and I always have sizeOfHeader in my all packet types
Now consider the situation : one type of my packets use sizeHeader to determine one of its fields size <Suppose it has at least one variable size field and it's size should be determined by sizeOfHeader - <const : I asked it before for that case> but one of them doesn't use sizeOfHeader as a field to determine something else size

So I cannot put the size of the header in Packet Base <which contain all common fields of my packets> because one of my packets enforce me to declear sizeOfHeader  in ACN and one of them enforces me to declare it in ASN 
Final result is that I should put sizeOfHeader in children <in all of my packets body> and this is not well organized

I think if you don't enforce fields that determine size for other fields must declare in ACN and make their declaration  available in ASN also it could solve this problem

**Reply by maxime-esa** [reply]
Regarding your first question, whenever you need to apply an offset you have to use a `mapping-function`, providing the computation of the field in a C/Ada external function.

I don't fully understand the other points, can you provide concrete examples?

**Comment by ErfanBabakhani** [comment]
```
Pdu DEFINITIONS AUTOMATIC TAGS ::= BEGIN

AddressSize ::= INTEGER (0..7)

HeaderSize ::= INTEGER (0..31)

Packet ::= SEQUENCE {
    --Common fields
    field1 INTEGER(0..255) ,
    --Common fields    
    child         Child
}

SrcAndDesAdd ::= OCTET STRING (SIZE (1..7))

Child ::= CHOICE {
    first          First,
    second         Second
}

First ::= SEQUENCE {  --in this PDU HeaderSize is always const excluding SrcAndDesAdd <We don't consider SrcAndDesAdd>
                                   --Because nobodies size depends on sizeHeader we declare it in ASN
    --Size of Address Field 3 bits: must declare in ACN
    sizeHeader     HeaderSize,
    srcAndDesAddress SrcAndDesAdd,
    crcHeaderr     CrcHeader
}

CrcHeader ::= OCTET STRING (SIZE (2))

SegmentCrc ::= OCTET STRING (SIZE (4))

Second ::= SEQUENCE {  --In this case size of header is var + consult so we declare it in ACN and by using the mapping function 
                                           add an offset for it
    --Size of Address Field 3 bits: must declare in ACN
    --Size of Header Field 5 bits: must declare in ACN
    srcAndDesAddress SrcAndDesAdd,
    var            Var,
    crcHeaderr     CrcHeader
}

Var ::= OCTET STRING (SIZE (0..16))

END
```
This is situation I was talking about

Also I have a new question 
How can I use postfunction property to calculate CrcHeader ?
